Coin Voting ZIP (#853)
zcash/zips: zip-0340.md

- The election parameters are immutable. If the EB
wants to make a modification, another election MUST be created

### Voter (ZH)

Anyone who has *spending authority* of ZEC 
(either transparent or shielded) can 
participate.

- Ballots MUST have a ZEC value
- Ballots MUST have a message in the format defined by the Election
- Ballots MUST include a proof of spending authority for the
amount specified, valid at the snapshot height
- Ballots MUST not prevent the ZH from spending the coins in a transaction
- Coins spent before the snapshot height MUST not be eligible for voting
- Voting MUST not disclose information that would allow someone 
to use the coins themselves (in a transaction or an election)
- Ballots MUST follow the format specified in the section Ballot
- ZEC held in transparent t3 addresses CANNOT be used. They introduce
most complexity and are not used much

Notes:
- Having coins on an exchange such as Binance, Bitfinex, etc.
does not allow you to vote because the coins are under their custody
- The EB MAY set special rules for the election. For example,
there MAY be a fee for casting a vote or a minimum ZEC requirement

> ## Privacy

Reviewer: This section currently describes how privacy issues could be solved. The ZIP will eventually instead need to address how privacy issues _are_ solved.

Most of this content should be moved to a Security and Privacy Considerations section, that will evolve as the ZIP draft is updated to address privacy issues. The only paragraph that should remain here is the one that includes conformance requirements.

If there are open questions about how to solve concerns that are given in the Security and Privacy Considerations section, add an Open Questions section outlining them (that can become a topic for discussion).

:::danger
If the coins used for voting for a given election are not moved
before voting for another election, it is possible to link
the votes.
:::

1. Transparent inputs are in clear
1. Shielded inputs reveal their nullifier

The total value is also revealed since it is the weight
of the vote.

> To mitigate this issue, the user SHOULD be offered to move
> their coins after a vote. However, if multiple elections
> are carried out concurrently with similar snapshot
> heights, it will be impossible for a Voter to participate
> in both without linking their votes.

Reviewer: This is the one paragraph that should stay in the Specification section.

Also, conformance requirements should be specific about what parties they place requirements upon:
```suggestion
Wallets SHOULD offer to move their user's coins after a vote. However, if
multiple elections are carried out concurrently with similar snapshot heights,
it will be impossible for a Voter to participate in both without linking their
votes.
```

    bytes nf_root = 10;
}

message ProofDomain {
    bytes domain = 1;
}

message Proof {
    bytes cv = 1;
    bytes nullifier = 2;
    bytes rk = 3;
    bytes proof = 4;
}

```

### Ballot Hash / Sighash

For the signatures, the sighash is the blake2b-256
hash of the serialization of the ballot data (in protobuf 3)
with personalization string `Zcash_WCV___Hash`.

`hash = Blake2b-256("Zcash_WCVSigHash", BallotData)`.

## Message Encryption

The EB MUST sign the ballot sighash with a ed25519 signing key and return
a receipt as a bech32m encoded signature with hrp `pob-receipt-`

The Voter MUST add a message in the format specified by the EB.
The message is encrypted into EncMessage.
- epk is a 32 byte public key on x25519
- the Voter reveals his own ed25519 signing key after the vote closes
- the EB converts the key to a x25519 secret key
- the EB uses DH to calculate a shared key KA
- SymKey = Blake2b-256(KA, perso="Zcash_ProofBlnce")
- SymKey is used with Chacha20-Poly1305 to decrypt the message
- The tag MUST match the decrypted tag

Rationale: This design allows the use of ed25519 SSH keys for both Voter and EB.

:::warning
The Voter SHOULD not reuse the signing key!
:::

> 

Reviewer: This is where the Security and Privacy considerations section should go:
```suggestion

# Security and privacy considerations

TBD
```
